File: e_DSA/b_DataStructures/a_Array/F_ThreeSum/Solution2.java
// 15. 3Sum => https://leetcode.com/problems/3sum/description/

package E_DSA.B_DataStructures.A_Array.F_ThreeSum;

public class Solution2 {
    

    static void threeSum(int[] nums) {
        int n= nums.length;

        for(int a=0; a<n; a++) {
            for(int b=a+1; b<n; b++) {
                for(int c=b+1; c<n; c++) {
                    if(nums[a]+nums[b]+nums[c]==0) {
                        System.out.println(nums[a] + " " + nums[b] + " " + nums[c]);
                    }
                }
            }
        }        
    }

    public static void main(String[] args) {
        int[] arr = {-1,0,1,2,-1,-4};

        threeSum(arr);
    }
}

File: DSA/DataStructures/Array/5. ThreeSum/Solution2.js
// https://leetcode.com/problems/3sum/description/
 var threeSum = function(nums) {

    nums.sort((x,y)=> x-y)
      
    let triplets = []
    for(let i=0; i<nums.length-2; i++) {

            if(i==0 || (i>0 && nums[i]!=nums[i-1])) {

                let j=i+1
                let k=nums.length-1
                let target=-nums[i]

                while(j<k) {

                    let sum = nums[j] + nums[k]

                    if(target == sum) {

                        triplets.push([nums[i], nums[j], nums[k]])

                        while(j<k && nums[j]==nums[j+1] ) j++
                        while(j<k && nums[k]==nums[k-1] ) k--
                         j++ 
                         k--
                    }

                    else if(sum>target){
                        while(j<k && nums[k]==nums[k-1] ) k--
                        k--
                    } 
                    else {
                        while(j<k && nums[j]==nums[j+1] ) j++
                        j++ 
                    }
                }  
                
            }
        }
        return triplets
    }